pyccel/parser/syntax/headers.py

```diff
@@ -59,15 +61,20 @@ def expr(self):
 class TemplateStmt(BasicStmt):
     """Base class representing a  template in the grammar."""
     def __init__(self, **kwargs):
-        self.dtype  = kwargs.pop('dtype')
+        self.dtypes  = kwargs.pop('dtypes')
         self.name   = kwargs.pop('name')
         BasicStmt.__init__(self)
 
     @property
     def expr(self):
-        l = [i.expr for i in self.dtype]
-        return Template(self.name, l)
+        if any(isinstance(d_type, FuncType) for d_type in self.dtypes):
+            msg = 'Functions in a template are not supported yet'
+            errors.report(msg,
+                        severity='fatal')
 
+        dtypes = tuple(dict(d_type) for d_type in {tuple(t.expr.items())\
+            for t in self.dtypes})
```

Comment:
I am not clear on what this does. You have:
```
tmp_dict = {tuple(t.expr.items())  for t in self.dtypes}
dtypes = tuple( dict(d_type) for d_type in tmp_dict )
```
So `tmp_dict` should be a dictionary combining the t.expr dictionaries. But then you iterate over it. Which means that you just use the keys. I am not sure how you create a dictionary from that. I am surprised this works!

Can you simplify this expression?

Reply:
`tmp_dict` in this case could be `{(('datatype', 'int'), ('precision', 4)), (('datatype', 'int'), ('precision', 8))}`, this is in order to remove any duplicated tuples that represents the types.
and then we convert the tuples back to dictionaries, `dict((('datatype', 'int'), ('precision', 4)))` would become `{'datatype': 'int', 'precision': 4}`. Do you think this is not a suitable approach ?

Reply:
Ok. That makes sense. It sounds suitable I just don't think it's very readable. Would you mind breaking it into 2 lines and giving the first line a variable with a name that explains what is happening. E.g:
```
possible_dtype_prec = {tuple(t.expr.items())  for t in self.dtypes}
dtypes = tuple( dict(d_type) for d_type in possible_dtype_prec )
```